### PR TITLE
Add symfony 3.0 support

### PR DESCRIPTION
**Pull request by Ener-Getick**
This PR adds symfony 3.0 support.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -14,11 +14,17 @@
             "homepage": "https://github.com/nelmio/NelmioApiDocBundle/contributors"
         }
     ],
+    "repositories": [
+        {
+            "type": "vcs",
+            "url": "https://github.com/Ener-Getick/DunglasApiBundle.git"
+        }
+    ],
```
Comment (Ener-Getick) [inline composer.json]: Must be removed once https://github.com/dunglas/DunglasApiBundle/pull/323 is merged.

File: Parser/FormTypeParser.php [hunk 2]
```diff
@@ -94,7 +94,7 @@ public function parse(array $item)
         $type    = $item['class'];
         $options = $item['options'];
 
-        if ($this->implementsType($type)) {
+        if (!method_exists('Symfony\Component\Form\AbstractType', 'getBlockPrefix') && $this->implementsType($type)) {
             $type = $this->getTypeInstance($type);
```
Comment (Ener-Getick) [inline Parser/FormTypeParser.php]: How can we replace this ? The support of ``FormTypeInterface`` is removed in SF 3.0 (see https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/FormFactoryInterface.php#L24).

**Comment by Ener-Getick** [comment]
ping @willdurand

**Comment by weaverryan** [comment]
@Ener-Getick This is GREAT work, but your tests are failing:

* 2.8 is failing due to Composer
* 3.0 is failing due to some deprecated things not being fixed yet - for example the `@` symbols in https://github.com/nelmio/NelmioApiDocBundle/blob/master/Tests/Fixtures/app/config/routing.yml aren't all quoted yet.

Could you take a look at the tests please?

Thanks!

**Comment by Ener-Getick** [comment]
I will try to fix the remaining deprecations tomorrow, I didn't localize all of them actually.

**Comment by Ener-Getick** [comment]
@xabbuh @fabpot do you know why there is no stable version of ``symfony/form`` on [packagist](https://packagist.org/packages/symfony/form) ?

**Comment by xabbuh** [comment]
Not sure what you mean. I see lots of stable versions.

**Comment by Ener-Getick** [comment]
Apparently this is fixed. A few minutes ago there were only the dev versions available.

**Comment by Ener-Getick** [comment]
I fixed several errors but the behavior of the symfony forms seems to be changed in sf 3.0 and so the NelmioApiDocBundle behavior ... :/

**Comment by weaverryan** [comment]
@Ener-Getick why did you close this? We can find a solution for supporting the form behavior across multiple versions - SonataAdminBundle also had to support this problem. I would love for you to continue your nice work to completion!

**Comment by Ener-Getick** [comment]
I didn't have any response so I thought you abandon. I'm not really involved in symfony core and in symfony/form so I don't know how we can fix this and why this behavior changed :-/

**Comment by xabbuh** [comment]
I'll try to help you with this after SymfonyCon (I.e. next week). Please ping me then if I forget about it.

**Comment by weaverryan** [comment]
:+1: @xabbuh 

And @Ener-Getick what are the problems that you're seeing? I noticed that you have the `LegacyFormHelper`, which is a nice way to handle the `text` -> `TextType` style of changes.

**Comment by Ener-Getick** [comment]
Thanks @xabbuh :-)

@weaverryan the output of the formatters has changed ([an example](https://travis-ci.org/nelmio/NelmioApiDocBundle/jobs/93953057#L489)) I don't know why... I'll try to search again this week. 

But I'm wondering if it wouldn't be easier to split this PR because it introduces a lot of changes to test at the same time...

**Comment by weaverryan** [comment]
Hmm, yea, I don't know about the formatting change - that is strange.

If it's easier for you to split across multiple PR's and get things working, you could do that. But I don't think it's too hard for a reviewer to review.

**Comment by Ener-Getick** [comment]
The problem is that I've made too many changes and I don't know when failures appeared... I'll create a new PR for everything already stable.

**Comment by Ener-Getick** [comment]
There are already several errors with symfony 2.8 ([see the build](https://travis-ci.org/nelmio/NelmioApiDocBundle/jobs/94006163)). I'll try to understand why this tests fail.

File: Tests/Fixtures/Form/ImprovedTestType.php [hunk 3]
```diff
@@ -24,18 +25,22 @@ class ImprovedTestType extends AbstractType
...
+            ->add('dt2', $datetimeType, array('date_format' => 'M/d/y'))
+            ->add('dt3', $datetimeType, array('widget' => 'single_text', 'format' => 'M/d/y H:i:s'))
+            ->add('dt4', $datetimeType, array('date_format' => \IntlDateFormatter::MEDIUM))
+            ->add('dt5', $datetimeType, array('format' => 'M/d/y H:i:s'))
+            ->add('d1', $dateType, array('format' => \IntlDateFormatter::MEDIUM))
+            ->add('d2', $dateType, array('format' => 'd-M-y'))
+            ->add('c1', $choiceType, array('choices' => array('m' => 'Male', 'f' => 'Female'), 'choices_as_values' => true))
+            ->add('c2', $choiceType, array('choices' => array('m' => 'Male', 'f' => 'Female'), 'choices_as_values' => true, 'multiple' => true))
+            ->add('c3', $choiceType, array('choices' => array()))
+            ->add('c4', $choiceType, array('choices' => array('foo' => 'bar', 'bazgroup' => array('baz' => 'Buzz')), 'choices_as_values' => true))
```
Comment (Ener-Getick) [inline Tests/Fixtures/Form/ImprovedTestType.php]: The users will have to update their form with ``choices_as_values`` if they want to use symfony 2.8 :/

**Comment by Ener-Getick** [comment]
The last commit speeds up travis of ~50% !
[before](https://travis-ci.org/nelmio/NelmioApiDocBundle/jobs/94149955) / [after](https://travis-ci.org/nelmio/NelmioApiDocBundle/jobs/94150647)

**Comment by Ener-Getick** [comment]
The remaining errors seem unrelated to this PR. Needs review

File: Parser/FormTypeParser.php [hunk 4]
```diff
@@ -94,11 +144,24 @@ public function parse(array $item)
...
+        if(!isset($form)) {
+            if (!LegacyFormHelper::hasBCBreaks() && $this->implementsType($type)) {
+                $type = $this->getTypeInstance($type);
+                $form = $this->formFactory->create($type, null, $options);
+            } else {
+                throw new \InvalidArgumentException('Unsupported form type class.');
+            }
+        }
 
         $name = array_key_exists('name', $item) ? $item['name'] : $form->getName();
```
Comment (luispabon) [inline Parser/FormTypeParser.php]: Form::getName has been removed from Symfony 3, use getBlockPrefix instead

Comment (Ener-Getick) [inline Parser/FormTypeParser.php]: Thanks, this line is apparently not covered by the tests...

**Comment by weaverryan** [comment]
The `asset()` function failure on 3.0 should be fixed by adding `asset: ~` into `default.yml` under the framework key: https://github.com/nelmio/NelmioApiDocBundle/blob/master/Tests/Fixtures/app/config/default.yml

But, that will also break the tests in 2.7 and lower. So in practice, we may need to add it in [AppKernel](https://github.com/nelmio/NelmioApiDocBundle/blob/master/Tests/Fixtures/app/AppKernel.php#L55), detect the Symfony version, and only add the asset key (this can be done via a closure) for 2.8 and higher.

Nice job btw!

**Comment by Ener-Getick** [comment]
Fixed, thanks @weaverryan

**Comment by Ener-Getick** [comment]
Should I rebase this PR ?

**Comment by weaverryan** [comment]
@Ener-Getick It's not up to me, and I don't care too much. But yea, if you could squash some of the commits where you're fixing many things in a row, that might make it a cleaner merge.

So this is ready for review and merge, right?

**Comment by Ener-Getick** [comment]
Yep I don't see anything else to change.

**Comment by Ener-Getick** [comment]
There's a deprecation on symfony 2.8.* probably caused by a third party. I'll search which one.

**Comment by Ener-Getick** [comment]
Apparently ``NumericNodeDefinition::cannotBeEmpty()`` has been deprecated in 2.8 but is still accessible in 3.0 thanks to its parent (see [here](https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Config/Definition/Builder/NumericNodeDefinition.php#L67)). 

This deprecation is triggered in ``dunglas/api-bundle`` but I do not understand why this function has been deprecated...

**Comment by Ener-Getick** [comment]
Hum apparently, this function is supposed to [throw an exception](https://github.com/symfony/symfony/pull/15170#issuecomment-119538615) in 3.0 ... Probably an oversight.

ping @dunglas the deprecation is [here](https://github.com/dunglas/DunglasApiBundle/blob/1.0/DependencyInjection/Configuration.php#L59)

**Comment by rande** [comment]
Hope this PR will be merged soon ;) It is a bloker for the sonata project!

**Comment by dunglas** [comment]
@Ener-Getick I merged your PR on API Bundle. Thanks!

**Comment by weaverryan** [comment]
@dunglas would it be possible to patch that and tag a 3.0-compat 1.0.2 release?

**Comment by dunglas** [comment]
@weaverryan As I merged the @Ener-Getick PR it should work right now because NelmioApiDoc depends only of the `@dev` version of the bundle: https://github.com/nelmio/NelmioApiDocBundle/blob/master/composer.json#L41

**Comment by willdurand** [comment]
Thanks everyone, because of SymfonyCon I don't know when I'll be able to merge this PR..

@rande got it. I'm going to change the initial roadmap so that this will be available in the upcoming minor version.

**Comment by Ener-Getick** [comment]
@willdurand no problem take your time ☺
@rande you can use my repository temporally to work on symfony 3.0 support ;-)

**Comment by willdurand** [comment]
Oh, one important thing though... THANK You so much for your work @Ener-Getick!

**Comment by willdurand** [comment]
mmh it looks like there is one remaining deprecation notice with 2.8:

```
The Symfony\Component\Config\Definition\Builder\NumericNodeDefinition::cannotBeEmpty method is deprecated since version 2.8 and will be removed in 3.0: 1x
    1x in DumpCommandTest::testDumpWithViewOption from NelmioApiDocBundle\Tests\Command
```

**Comment by Ener-Getick** [comment]
You're welcome ;-)

And yeah, see [this](https://github.com/nelmio/NelmioApiDocBundle/pull/752#issuecomment-161342644). Can you relaunch Travis? Otherwise I'll do it tonight.

**Comment by willdurand** [comment]
@Ener-Getick that's what I did (triggering travis-ci again), but it is still failing...

**Comment by Ener-Getick** [comment]
Yeah just see that dunglas/api-bundle is under 2.0.x-dev

**Comment by willdurand** [comment]
oh yes... @dunglas can we safely rely on `2.0@dev`?

**Comment by dunglas** [comment]
No we can't but AFAIK the @Ener-Getick patch (thx for the good work) has been merged in `1.0@dev`.

**Comment by Ener-Getick** [comment]
I just discovered that packagist detects the branches named ``\d.\d`` the problem is that the last dev release is ``v1.1.0-beta.1`` and ``1.0@dev < v1.1.0-beta.1``.

@dunglas could you rename your branch to ``1.1`` ?

**Comment by dunglas** [comment]
I've just created a 1.1 branch (it's cleaner I think). It should work when #764 is merged.

**Comment by dunglas** [comment]
In fact renaming the `1.0` branch to `1.x` do the trick. Can you run Travis again?

**Comment by Ener-Getick** [comment]
It works thanks @dunglas

**Comment by willdurand** [comment]
awesome, thanks!

**Comment by dunglas** [comment]
Thanks a lot @Ener-Getick

**Comment by rande** [comment]
GG!

**Comment by luispabon** [comment]
Thank you :+1:

-- 



1 Neal Street London WC2H 9QL

www.rightster.com | @Rightster <https://twitter.com/rightster> | LinkedIn 
<https://www.linkedin.com/company/rightster>


Rightster Limited (incorporated in England with company number 07634543) 
has its registered office at 1 Neal Street London WC2H 9QL. Rightster 
Limited is a subsidiary of Rightster Group plc (incorporated in England and 
Wales with company number 08754680).

**Comment by Ener-Getick** [comment]
Awesome thank you all for your help

**Comment by teohhanhui** [comment]
I get this deprecation notice on Symfony 2.8:

> The alias option of the form.type_extension tag is deprecated since version 2.8 and will be removed in 3.0. Use the extended_type option instead.

This is because of the "alias" attribute here: https://github.com/nelmio/NelmioApiDocBundle/blob/1df3ddf49f4569b0b5f84780373671ed96e6025b/Resources/config/services.xml#L40

**Comment by Ener-Getick** [comment]
Hum this deprecation has probably been added later. I'll fix it this afternoon.

**Comment by Ener-Getick** [comment]
Which version of the bundle do you use ? This should not trigger a deprecation see [here](https://github.com/symfony/symfony/blob/2.8/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/FormPass.php#L64)

**Comment by teohhanhui** [comment]
@Ener-Getick You're right. It must be caused by another bundle then... (The way deprecation warnings are logged is pretty unhelpful.)

**Comment by weaverryan** [comment]
@teohhanhui you can open an issue in symfony/symfony if a deprecation warning isn't clear enough - it's pretty common that many can be improved to contain more information (in this case, if the service id were in the message, that would help a lot).

**Comment by willdurand** [comment]
@teohhanhui [v2.11.1](https://github.com/nelmio/NelmioApiDocBundle/releases/tag/2.11.1) is released and contains @Ener-Getick's fix.

**Comment by Ener-Getick** [comment]
I forgot to specify, it's a fix for a deprecation triggered in the tests not in production :x

**Comment by willdurand** [comment]
@Ener-Getick yep, but it documents how to fix this deprecation :)

**Comment by teohhanhui** [comment]
Uhh, yes. I rechecked and it turns out I was using `2.10.3`. The deprecation notice I was getting was already fixed by this PR (in `2.11.0`). Silly me!

**Comment by OskarStark** [comment]
:+1:

**Pull request merged**